app/src/main/java/org/schabi/newpipe/player/BackgroundPlayer.java

```diff
@@ -633,5 +641,48 @@ public void onCompleted() {
             updateNotification(R.drawable.ic_replay_white);
             lockManager.releaseWifiAndCpu();
         }
+
+        @Override
+        public void setTimer(int hourOfDay, int minute) {
+            long time = ((hourOfDay * 60) + minute) * 60 * 1000;
+            if (time == 0) {
+                Toast.makeText(getApplicationContext(), getString(R.string.timer_select_time_message), Toast.LENGTH_SHORT).show();
+                return;
+            }
+            timer.cancel();
+            timer.purge();
+            timerHour = hourOfDay;
+            timerMinute = minute;
+            timer = new Timer();
+            timer.schedule(new TimerTask() {
+                @Override
+                public void run() {
+                    new Handler(Looper.getMainLooper()).post(() -> {
+                        if (DEBUG) Log.d(TAG, getString(R.string.log_stopping_the_player));
+                        onPause();
+                    });
+
+                }
+            }, time);
+            Date d = new Date();
+            d.setTime(System.currentTimeMillis() + time);
+            Toast.makeText(getApplicationContext(), String.format(getString(R.string.player_stop_message), d.toString()), Toast.LENGTH_LONG).show();
+        }
+
+        @Override
+        public void cancelTimer() {
+            timer.cancel();
+            timer.purge();
+        }
+
+        @Override
+        public int getHourOfDay() {
+            return this.timerHour;
+        }
+
+        @Override
+        public int getMinutes() {
+            return this.timerMinute;
+        }
```

Comment:
Move this whole block to BasePlayer, replacing the base functions with unsupported operations.

Reply:
Done